# KING COUNTY REAL ESTATE ANALYSIS

Please fill out:
* Student name: Ian Musau, Mathew Karani, Jacinta Chepkemoi, Christine Malinga and Tabitha Berum. GROUP 2
* Student pace: Full time, Hybrid
* Scheduled project review date/time: 02/05/2024
* Instructor name: Nikita Njoroge
* Blog post URL:


## Overview


The real estate sector encompasses a diverse range of activities involving the acquisition, development, management, and transaction of properties, including residential, commercial, and investment real estate. The urban real estate investment business model is undergoing a fundamental overhaul attributed to digitization and a growing market for smart and environmentally demanding buildings.
Data analysis tools can be used to extract valuable information for important decision-making such as house price predictions.

 ## Business Undersatnding



A client has a property in King County, WA that they are looking to sell. Having no prior knowledge or experience with the real estate industry, he has chosen to approach our agency so as not miss out on the chance of maximizing their profit potential. The main concern is the potential renovation to focus on that would have the least effect in production price and still increase property value.
Data being analyzed is housing data from King County, WA in order to identify what opportunities there are to increase the sales price of the home, and by how much.

This analysis will focus on the following factors that majorly affect the real estate industry to achieve the client's request:
i) Market Trends: This help real estate agents, investors, and other stakeholders make well-informed decisions about the purchase, sale, and investment strategies of real estate, the dataset provides insights into pricing dynamics, market trends, and property features.
ii) Customer Preferences: Understanding of consumer preferences and market demand through analysis of bedroom and bathroom numbers, property condition, view ratings makes targeted markeing and property customization easier and more scalable so as to suit all their clients' expectations.
iii) Investment Opportunities: By analyzing variables including property size, location, condition, and market trends, investors in real estate can find properties that are cheap, evaluate possible returns, and optimize their investment portfolios.
iv) Competitive Analysis: Real estate agents can benchmark against rivals, spot market gaps, and more by comparing property prices, sizes, amenities, and location features.
and set themselves out from the competition to obtain a market advantage.

## Data Understanding

The dataset "house_data" contains information about real estate properties, such as property IDs, price ranges, dates of sales, number of bedrooms and bathrooms, living space and lot sizes, floor counts, waterfront access, view ratings, property condition and grade, years of construction and renovation, geographic coordinates, and information about competitive neighborhoods.

Gaining an understanding of the differences included in each characteristic and how they are distributed is essential for doing important analyses, such as identifying the factors influencing property values, assessing neighborhood characteristics, and predicting market trends. Furthermore, it will be crucial to address data quality issues like missing values, outliers, and inconsistent data types in order to ensure the accuracy and reliability of any studies or models based on this dataset.

In [59]:
# import relevant libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [60]:


class DataSourcing:
  def __init__(self,dataframe):
    self.original = dataframe
    self.dataframe = dataframe
  
  def give_info(self):
    message =  f"""
    ----------------------------------------------------------------------->
    DESCRIPTION OF THE DATAFRAME IN QUESTION:
    ----------------------------------------------------------------------->
    
    Dataframe information => {self.dataframe.info()}
    ------------------------------------------------------------------------------------------------------------------------->
    
    Dataframe shape => {self.dataframe.shape[0]} rows, {self.dataframe.shape[1]} columns
    ------------------------------------------------------------------------------------------------------------------------->    
    
    There are {len(self.dataframe.columns)} columns, namely: {self.dataframe.columns}.  
    ------------------------------------------------------------------------------------------------------------------------->
        
    The first 5 records in the dataframe are seen here:
    ------------------------------------------------------------------------------------------------------------------------->
    {self.dataframe.head()}
    ------------------------------------------------------------------------------------------------------------------------->
       
    The last 5 records in the self.dataframe are as follows: 
    ------------------------------------------------------------------------------------------------------------------------->
    {self.dataframe.tail()}
    ------------------------------------------------------------------------------------------------------------------------->
    
    The descriptive statistics of the dataframe (mean,median, max, min, std) are as follows:
    ------------------------------------------------------------------------------------------------------------------------->
    {self.dataframe.describe()}
    ------------------------------------------------------------------------------------------------------------------------->
    """
    print (message)
  
  def null_count(self):
    return self.dataframe.isnull().sum()
  
  def unique_count(self):
    return self.dataframe.nunique()
  
  def unique_per_column(self):
    print("<----- UNIQUE VALUES IN EACH COLUMN ----->")
    for col in self.dataframe.columns:
      print(f"{col} ({len(self.dataframe[col].unique())} unique)\n {sorted(self.dataframe[col].unique())}")
      print()
    print("<----- END OF UNIQUE VALUES IN EACH COLUMN ----->")
    return
  
  
 
 




Load the CSV dataset

In [61]:
# Call the load_data() function to load the data
self = pd.read_csv('data\kc_house_data.csv')


Now we can start the data understanding and pass the dataframe to our DataSourcing class

In [62]:
house_data = DataSourcing(dataframe=self)
house_data.give_info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

The insights derived from the give_info() method provide essential details about the dataset, contributing to a comprehensive understanding:

This data set has 21 columns namely: 
id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15.

Missing Values:
waterfront, view, and yr_renovated columns have missing values. This can be seen in the count row of the descriptive statistics section.

Potential Data Quality Issues:
The bedrooms column has a maximum value of 33, which seems unusually high and might be an error or outlier.
The bathrooms column has a maximum value of 8, which could also be considered high and should be examined for outliers.
The yr_renovated column has a maximum value of 2015, which seems unusual as it's the same as the maximum value of the yr_built column. This suggests that some values in the yr_renovated column might represent the year built instead of renovation years.

Inconsistent Data Types:
Some columns, such as waterfront, view, condition, and grade, appear to have categorical data but are represented as objects (strings) instead of categorical data types.
The date column is represented as an object (string) but should be converted to a datetime data type for easier manipulation and analysis.

Potential Outliers:
Outliers may exist in numerical columns such as price, bedrooms, bathrooms, sqft_living, sqft_lot, sqft_above, sqft_basement, yr_built, yr_renovated, lat, long, sqft_living15, and sqft_lot15. Visualizing the distributions of these columns can help identify outliers.

Zipcode as Numeric:
The zipcode column is currently represented as an integer, but it might be more appropriate to treat it as a categorical variable since it represents different geographic areas.

Inconsistent Naming Conventions:
Some column names are in snake_case format (sqft_living, sqft_lot) while others are in camelCase (sqftLiving15, sqftLot15). It's best to use a consistent naming convention throughout the dataset.

In [63]:
# Check for unique data so as to avoid redundancy in our dataset
data_sourcing.unique_count()

id               21420
date               372
price             3622
bedrooms            12
bathrooms           29
sqft_living       1034
sqft_lot          9776
floors               6
waterfront           2
view                 5
condition            5
grade               11
sqft_above         942
sqft_basement      304
yr_built           116
yr_renovated        70
zipcode             70
lat               5033
long               751
sqft_living15      777
sqft_lot15        8682
dtype: int64

Upon examining the data, it appears to be authentic and devoid of anomalies for several reasons:

Variety in Property IDs: The dataset comprises 21,420 distinct IDs, each representing a unique property.
Diverse Sale Dates: A total of 372 unique sale dates are recorded, reflecting the temporal spread of property transactions.
Price Variation: With 3,622 unique price points, the dataset demonstrates a wide range of property prices.
Bedroom and Bathroom Distribution: There are 12 distinct bedroom counts and 29 unique bathroom counts, indicating diverse property configurations.
Living Space and Lot Sizes: The dataset encompasses 1,034 different living space sizes and 9,776 distinct lot sizes, showcasing the variability in property dimensions.
Floor Counts: Properties are characterized by 6 unique floor counts, illustrating differences in architectural design.
Waterfront Access: The "waterfront" column exhibits 2 unique values, indicating the presence or absence of waterfront access for properties.
View Ratings: Properties are rated on 5 unique view levels, reflecting varying scenic qualities.
Condition and Grade: The dataset includes 5 distinct property conditions and 11 unique grade categories, offering insights into property quality.
Basement and Above-Ground Area: Property basements and above-ground areas are characterized by 304 and 942 unique size values, respectively.
Construction and Renovation Years: There are 116 unique construction years and 70 distinct renovation years, providing historical context for properties.
Geographical Distribution: Properties are located in 70 unique zip codes, with 5,033 distinct latitude and 751 longitude values, showcasing geographical diversity.
Neighborhood Comparison: The dataset features 777 different living space sizes and 8,682 lot sizes for the 15 nearest neighbors, enabling neighborhood-specific analyses.

The dataset exhibits substantial diversity across various property attributes, including size, location, condition, and price. These unique counts offer valuable insights into the dataset's composition, facilitating informed analyses and decision-making processes within the real estate domain.